Renew LetsEncrypt certificates

iulianbarbu opened the pull request:
Closes #621 .

## Description

This PR adds functionality that enables the periodical renewal of  `LetsEncrypt` certificates. These certificates have a validity window of 90 days and is recommended to be renewed 30 days before expiration. The functionality will cover renewal for:

1) certificates emitted for a custom domain project when requested through the admin gateway API `"/admin/acme/request/:project_name/:fqdn"` or through `shuttle-admin` CLI.
2) the gateway certificate, which is loaded when starting the gateway service or created based on a set of ACME credentials present on the local filesystem.

## WIP testing

1) End-to-end manual test of the certificate renewal with all shuttle components deployed locally. I have some issues with the shuttle local deployment that blocks me while testing. I need some more time to understand how to enable shuttle locally, with pebble deployed locally as CA.

2) Unit testing for the introduced business logic where it makes sense and is possible. Spent lots of time on 1) and I'll need to come back at this.

I welcome any suggestions on steps/guides/docs for how to enable the shuttle components locally to test the certificates renewal.

## Documentation

Haven't seen much documentation on how to test the admin functionality related to the certificates requesting, so I hope I'll be able to compile something worth mentioning after finishing this.

## Review wise

Besides code review, I am curious at the beginning to understand if this PR covers what's asked from #621 functionality-wise. I would be happy to follow up with adjustments code-wise if there are suggestions for improvement. Also, in parallel, I'll try to come back with information on the testing aspect.

iulianbarbu commented:
I am blocked by the fact that my pebble local deployment finalizes an order but it's not returning the certificate url, needed by the ACME client to download the certificate. Can't say what's the order status after finalizing, I'll need to debug by working through the `instance-acme` dependency code.

So far, the pebble CA log looks like this:

```
Pebble 2023/02/20 17:55:06 ACME directory available at: https://0.0.0.0:14000/dir
Pebble 2023/02/20 17:56:55 GET /dir -> calling handler()
Pebble 2023/02/20 17:56:55 HEAD /nonce-plz -> calling handler()
Pebble 2023/02/20 17:56:55 POST /sign-me-up -> calling handler()
Pebble 2023/02/20 17:56:55 There are now 1 accounts in memory
Pebble 2023/02/20 17:57:19 HEAD /nonce-plz -> calling handler()
Pebble 2023/02/20 17:57:19 POST /order-plz -> calling handler()
Pebble 2023/02/20 17:57:19 There are now 1 authorizations in the db
Pebble 2023/02/20 17:57:19 Added order "niAlrdD77-VhHeU6raADNxxxYuZSjCqQujLmUTCmirI" to the db
Pebble 2023/02/20 17:57:19 There are now 1 orders in the db
Pebble 2023/02/20 17:57:19 POST /authZ/ -> calling handler()
Pebble 2023/02/20 17:57:19 POST /chalZ/ -> calling handler()
Pebble 2023/02/20 17:57:19 Pulled a task from the Tasks queue: &va.vaTask{Identifier:acme.Identifier{Type:"dns", Value:"hello-world-rocket-app.unstable.shuttleapp.rs"}, Challenge:(*core.Challenge)(0xc0001c8280), Account:(*core.Account)(0xc0000785a0)}
Pebble 2023/02/20 17:57:19 Starting 3 validations.
Pebble 2023/02/20 17:57:19 Sleeping for 3s seconds before validating
Pebble 2023/02/20 17:57:19 Sleeping for 2s seconds before validating
Pebble 2023/02/20 17:57:19 Sleeping for 2s seconds before validating
Pebble 2023/02/20 17:57:20 POST /my-order/ -> calling handler()
Pebble 2023/02/20 17:57:20 POST /my-order/ -> calling handler()
Pebble 2023/02/20 17:57:21 POST /my-order/ -> calling handler()
Pebble 2023/02/20 17:57:21 Attempting to validate w/ HTTP: http://hello-world-rocket-app.unstable.shuttleapp.rs:7999/.well-known/acme-challenge/trr79hstEGoLoSkdJfdQWxAZq7Ufl1BDG4PmKzsMVks
Pebble 2023/02/20 17:57:21 Attempting to validate w/ HTTP: http://hello-world-rocket-app.unstable.shuttleapp.rs:7999/.well-known/acme-challenge/trr79hstEGoLoSkdJfdQWxAZq7Ufl1BDG4PmKzsMVks
Pebble 2023/02/20 17:57:22 Attempting to validate w/ HTTP: http://hello-world-rocket-app.unstable.shuttleapp.rs:7999/.well-known/acme-challenge/trr79hstEGoLoSkdJfdQWxAZq7Ufl1BDG4PmKzsMVks
Pebble 2023/02/20 17:57:22 authz UFddOjhDm8lo3tgfjdAvU4pJYo8l8iIgTYFrIE4IsvE set VALID by completed challenge -lhkvnOJCIG4rRoDj6b-ijmsfUD_GiKAXKkl1ekwmo8
Pebble 2023/02/20 17:57:23 POST /my-order/ -> calling handler()
**Pebble 2023/02/20 17:57:23 POST /finalize-order/ -> calling handler()
Pebble 2023/02/20 17:57:23 Order niAlrdD77-VhHeU6raADNxxxYuZSjCqQujLmUTCmirI is fully authorized. Processing finalization
Pebble 2023/02/20 17:57:23 Issued certificate serial 05cf8f20a311617a for order niAlrdD77-VhHeU6raADNxxxYuZSjCqQujLmUTCmirI**
```

It seems that the order finalization succeeded with a certificate issued, but the finalization response doesn't contain the certificate URL, or the logs don't mention it sadly.

The gateway logs look like this:

<img width="713" alt="Screenshot 2023-02-20 at 18 27 22" src="https://user-images.githubusercontent.com/14218860/220158905-cb9d0b1c-b1ec-4d5e-8140-65b40018ed46.png">

The place where this fails is in `instant-acme` finalize order method:
![Screenshot 2023-02-20 at 18 28 47](https://user-images.githubusercontent.com/14218860/220159258-54382b23-f948-42a0-85a2-ac0c40ecbc81.png)

brokad commented on gateway/src/main.rs:
```diff
@@ -237,46 +254,93 @@ async fn init(db: SqlitePool, args: InitArgs) -> io::Result<()> {
+    resolver: Arc<GatewayCertResolver>,
+) {
+    loop {
+        tokio::time::sleep(Duration::from_secs(240)).await;
```
Nit: Why not sleep/wait until the first time when we actually want this to fire? Instead of `loop { ... }`?

iulianbarbu replied:
I think this will not be required anymore given we can not properly automate the gateway certificate renewal without the TXT record insertion for the DNS challenge. I can open an issue to track this issue and handle it separately, is that ok?

brokad commented on gateway/src/main.rs:
```diff
@@ -237,46 +254,93 @@ async fn init(db: SqlitePool, args: InitArgs) -> io::Result<()> {
+        let diff = x509_cert.validity().not_after.sub(ASN1Time::now()).unwrap();
+        if diff.whole_days() <= 30 {
+            let tls_path = fs.as_ref().join("ssl.pem");
+            let certs = create_gateway_certificate(
```
The strategy LGTM, but the only thing I'm concerned with is the challenge itself. Since this eventually calls `AcmeClient::create_certificate` with a challenge which is DNS-based, if the challenge key is different from the previous one this time round, then we'll need manual intervention and update our TXT records for `shuttleapp.rs` at the same time. Or give this service more AWS permissions so that it can update them on its own.

If we decide to go manual intervention (which is definitely easier), then better keep this behind an admin CLI subcommand that we fire ourselves. We can then "automate" it by having a simple (but separate) cron job that fires that admin subcommand once in a while, all we need this job to know about is what to update the TXT record to.

iulianbarbu replied:
Hah, you're right. The DNS challenge is a beast to secure [1]. This might need some dev work with AWS services, but I think before that we can start with introducing the support for requesting the gateway certificate through an admin command, on-demand, and leave the DNS challenge completion (required for certificate renewal) to be handled in a separate PR. What do you think?

[1] https://www.eff.org/deeplinks/2018/02/technical-deep-dive-securing-automation-acme-dns-challenge-validation

iulianbarbu replied:
I've implemented admin support to request for gateway certificate renewal (with DNS TXT record insertion) out of band in terms of API and cmdline. Let me know in case there is something else to address.

brokad commented on gateway/src/api/latest.rs:
```diff
@@ -362,6 +352,91 @@ async fn request_acme_certificate(
+                            .unwrap();
+                        // If current certificate validity less_or_eq than 30 days, attempt
+                        // renewal.
+                        if diff.whole_days() <= 30 {
```
If this is fired only through the admin command, why is there a check for expiration here? IOW, why don't we just expose the ability to (re)create certificates through the admin subcommand? And then whatever calls that command can do it on arriving close to expiration (and we gain the ability to renew them out of band should we want to).

For that, we could just expose an admin endpoint for listing custom certs maybe?

iulianbarbu replied:
For sure, keeping the automatic renewal out of the band should be cleaner. Then what is left is only the certificate creation, which is already contained by the `admin/acme/renew` API. The difference between this API and the `admin/acme/request` API is that the latter will additionally destroy and recreate the project. 

I will remove the task from the `admin/acme/renew` API and rely on an out-of-band automated renewal (cron job-like component for each project).

iulianbarbu replied:
Removed the automation for both gateway and custom domain certificates renewal. We'll have to rely on an out of band mechanism that calls into the shuttle-admin commands or gateway APIs. Let me know if you would like to change something or add more to this.

chesedo commented:
Hey @iulianbarbu, thanks for the PR :smile: 

> It seems that the order finalization succeeded with a certificate issued, but the finalization response doesn't contain the certificate URL, or the logs don't mention it sadly.

Gateway should create a trace event with "got authorization" (a few lines earlier). Iirc, this log will contain some challenge URL that can help with the debugging

iulianbarbu commented:
Hey @chesedo ! Thanks for taking a look. On the pebble side, the above log I sent contains this at the end, but haven't checked the corresponding trace events on the gateway side. I'll try again.

> Pebble 2023/02/20 17:57:23 POST /my-order/ -> calling handler()
**Pebble 2023/02/20 17:57:23 POST /finalize-order/ -> calling handler()
Pebble 2023/02/20 17:57:23 Order niAlrdD77-VhHeU6raADNxxxYuZSjCqQujLmUTCmirI is fully authorized. Processing finalization
Pebble 2023/02/20 17:57:23 Issued certificate serial 05cf8f20a311617a for order niAlrdD77-VhHeU6raADNxxxYuZSjCqQujLmUTCmirI**

iulianbarbu commented:
@brokad , I've addressed your comments here: https://github.com/shuttle-hq/shuttle/pull/641/commits/5ad20c0f77d9fa27424a376eab53e5aa93b611c3. Let me know in case you have other concerns. 

Other than this, I will take another stab at the end-to-end test and then unit testing + docs.

chesedo commented on admin/src/args.rs:
```diff
@@ -55,8 +55,34 @@ pub enum AcmeCommand {
+
+        /// Path to acme credentials file
+        /// This should have been created with `acme create-account`
+        #[arg(long, value_parser = load_credentials)]
+        credentials: serde_json::Value,
```
Wow, nice update with the `value_parser` :star_struck:

chesedo reviewed (approved):
Thanks for the updates @iulianbarbu

chesedo commented on gateway/src/acme.rs:
```diff
@@ -150,15 +142,27 @@ impl AcmeClient {
+                AcmeClientError::CertificateCreation
             })?;
+            retries -= 1;
+            sleep(Duration::new(1, 0)).await;
```
```suggestion
            sleep(Duration::from_secs(1)).await;
```
Just like how this duration time is clearer without having to look up the docs

iulianbarbu replied:
I am not sure what you mean

oddgrd replied:
Pieter's suggestion is just more readable since you don't need to look up what the args to `Duration::new` are, and it's how we do it elsewhere in the codebase.

iulianbarbu replied:
Ah, right. Didn't notice he change the method to reduce the args count. Makes sense, will do! Thanks guys! 🤘

iulianbarbu commented on Containerfile:
```diff
@@ -3,18 +3,21 @@ ARG RUSTUP_TOOLCHAIN
 WORKDIR /src
 COPY . .
-RUN find ${SRC_CRATES} \( -name "*.proto" -or -name "*.rs" -or -name "*.toml" -or -name "Cargo.lock" -or -name "README.md" -or -name "*.sql" \) -type f -exec install -D \{\} /build/\{\} \;
+RUN find ${SRC_CRATES} \( -name "*.proto" -or -name "*.rs" -or -name "*.toml" -or -name "Cargo.lock" -or -name "README.md" -or -name "*.sql" -or -name "*.pem" \) -type f -exec install -D \{\} /build/\{\} \;
```
Indeed. I think it's useful only for dev environment so I might guard it behind a dev check or remove it altogether. WDYT?

oddgrd commented on Makefile:
```diff
@@ -88,13 +94,14 @@ clean:
 
 images: shuttle-provisioner shuttle-deployer shuttle-gateway shuttle-auth postgres panamax otel
 
+# We must specify the platform tag so that
```
I think we can remove this comment, right?

oddgrd commented on admin/README.md:
Thank you so much for this guide! :heart_eyes:

oddgrd commented on gateway/Cargo.toml:
```diff
@@ -20,7 +20,7 @@ http = { workspace = true }
 # not great, but waiting for WebSocket changes to be merged
 hyper-reverse-proxy = { git = "https://github.com/chesedo/hyper-reverse-proxy", branch = "bug/host_header" }
-instant-acme = "0.1.1"
+instant-acme = "0.2.0"
```
Great job securing the upstream fix in instant-acme!

oddgrd commented on gateway/src/main.rs:
```diff
@@ -141,7 +140,7 @@ async fn start(db: SqlitePool, fs: PathBuf, args: StartArgs) -> io::Result<()> {
 
     let mut api_builder = ApiBuilder::new()
-        .with_service(Arc::clone(&gateway))
+        .with_service(Arc::<GatewayService>::clone(&gateway))
```
I see you did this in a couple of places, I'm curios why this is needed?

iulianbarbu replied:
I've developed for some time with 1.67.1 before I've changed to 1.65. At that time my vs code recommended me to include the turbofish operator or I couldn't compile the source code (can't remember which one). I blamed it on the 1.67.1 version at that time but I can not reproduce it anymore. I'll remove the turbofish operator because it compiles just fine without it.

oddgrd reviewed (approved):
Thanks for this Iulian, LGTM to me too! I left a few comments but mostly just nice things and a question or two :smile:

chesedo commented on Containerfile:
```diff
@@ -3,18 +3,22 @@ ARG RUSTUP_TOOLCHAIN
 WORKDIR /src
 COPY . .
 RUN find ${SRC_CRATES} \( -name "*.proto" -or -name "*.rs" -or -name "*.toml" -or -name "Cargo.lock" -or -name "README.md" -or -name "*.sql" \) -type f -exec install -D \{\} /build/\{\} \;
+RUN [ "$CARGO_PROFILE" != "release" ] && find ${SRC_CRATES} -name "*.pem" -type f -exec install -D \{\} /build/\{\} \;
```
I like this... well, I can't really think of a better way to do this :see_no_evil: 

We should just add a comment to explain the thinking here for our future selves

chesedo reviewed (approved):
Thanks for all the edits @iulianbarbu. LGTM
